content/2021-06-07-quarkus-func-payload-mapper.md

```diff
@@ -0,0 +1,308 @@
+```
+Now we should have everything set, so let's test it out.
+
+### Testing
+
+Let's create a matching app and device in Drogue cloud (in the future we can sync creation of resource in Drogue and Ditto using registry change events).
+
+```shell
+drg create app app_id
+drg create device --app app_id simple-thing --data '{"credentials": {"credentials":[{ "pass": "foobar" }]}}'
+```
+
+And we're ready to send some data
+
+```shell
+http --auth simple-thing@app_id:foobar --verify build/certs/endpoints/ca-bundle.pem POST https://$HTTP_ENDPOINT/v1/foo data_schema==ditto:test temp:=21.5 hum=51.0
+```
+
+Now we can check the state of our twin and make sure our changes are persisted
+
+```shell
+http --auth ditto:ditto http://$TWIN_API/api/2/things/app_id:simple-thing
+HTTP/1.1 200 OK
+...
+etag: "rev:2"
+response-required: false
+version: 2
+
+{
+    "features": {
+        "hum": {
+            "properties": {
+                "value": "51.0"
+            }
+        },
+        "temp": {
+            "properties": {
+                "value": 21.5
+            }
+        }
+    },
+    "policyId": "app_id:simple-thing",
+    "thingId": "app_id:simple-thing"
+}
+```
+
+So we can see that our sensor data went via HTTP endpoint to Knative Kafka channel and then through our payload converter function reached the digital twin platform.
+
+## Future
+
+We hope this post make you realize how Func in combination with Quarkus make a good platform for creating serverless function that have real applications in IoT cloud platforms. We'll strive to make the experience even better in the future. Here are some notes on the experience and future improvements.
+
+First on the Func side, the current function developer workflow is generally good. What would make it even better is making it more tied with the accompanied resources (sources and sequences in our case). We can do two things in that area. We can make `func` CLI manage (deploy/delete) custom yaml resources along with the service itself. But also, it'd be good if `func` would just be able to generate service yaml resource so we can add it to existing workflows (GitOps for example).
+
+On the Quarkus side there are a few small improvements that could be made. Even if the docs says, it requires Java 11, it actually sets Java 8 as default version. Also, the Funq implements its own Cloud Events abstractions, which are not as polished as official SDK. And using the official ones would be a better choice to stay current with the spec implementation in the long run. These are all small niggles that we're gonna feed back into the upstream community and try to improve upon.
+
+And of course, the Rust support for Func is on the way (still in early development). We want to make Rust first-class citizen for developing serverless functions and provide a great experience. And as we want to use them in Drogue cloud we want to help out with the effort. So stay tuned.
```

Comment:
In case you want to add links... 
https://github.com/boson-project/buildpacks/pull/95
https://github.com/boson-project/func/pull/376

Reply:
Added